UM-USElab/gradecraft-development
# added auto award and email notification functionality to LO model

### Status
**Ready**

### Description
Added the ability to have a learning objective unlockable condition that automatically awards a badge.
Also added email notification to the achievement of a learning Objective

### Related PRs
List related PRs against other branches:

### Todos
- [x] More user testing for emails and functionality -- Make sure students are emailed upon completion of a learning objective
- [x] Email notification saying that the user has completed the learning objective
- [x] QA / User testing. See below or ask me (Erik) if you have any questions
- [x] Add rspec tests

### Migrations
YES -- Made migration to add email_learning_objective_achieved to course_memberships & achieved to learning_objectives_cumulative_outcomes

### Steps to Test or Reproduce
Create or Edit a badge to have a learning objective as the unlockable condition. Underneath while editing check the 'Auto award on unlock' box. Then have a student achieve the learning objective. This results in the badge getting awarded and email notifications sent automatically.

In addition to the badge getting auto awarded, this PR also includes functionality around emailing a user when a learning objective is achieved for the first time. 

To ensure an email is sent upon the completion of a Learning Objective (LO) the email setting must be checked on /users/edit_profile (reachable by clicking on your name in the top right corner and clicking 'my account'). 

Now, the achievement of a LO should also email the student saying it has been achieved.

Things to make sure they work properly --
* Achievement of LO sends email, achieved by either a number of points achieved or assignments completed.  
* receiving a proficiencies for ONE assignment that achieves TWO learning objectives sends TWO emails
* Achieving a learning objective automatically awards a badge and sends appropriate amounts of emails -- Achieving LO by a grade should send grade email, LO achieved email, badge unlocked email, badge awarded email
* Unchecked 'email learning objective achieved' should not send an email upon the completion of a LO

### Impacted Areas in Application
Learning objective model. Email notifications. grade model

======================
Closes #4139 #4033
